### PR TITLE
Add v0 type mangling prefixing for simple types

### PR DESCRIPTION
**Pull request by CohenArthur**
This PR adds the generation of type prefixes for simple types, which are numeric types, booleans, chars, strings, empty tuples/unit types and placeholder types.

I'm unsure as to how to test this, even in the long run. There might be some shenanigans we can pull using an elf reader and regexes in order to compare ABI names with rustc.

The entire implementation of v0 name mangling is very large, so I thought I'd split it up in multiple PRs.

### CODE REVIEW

**Comment by dkm** [comment]
I guess you should use the `scan-assembler` (see https://gcc.gnu.org/onlinedocs/gccint/Final-Actions.html#Final-Actions ) to look for a specific symbol name somewhere ?

**Comment by CohenArthur** [comment]
Fantastic Marc, thanks a lot!!!! I didn't know this existed. I guess we can think about comparing the output to rustc later on but it would be quite easy with this

**Comment by dkm** [comment]
Yes. You can already check that you end up with the expected names in the assembly :).

**Review by philberty — APPROVED** [review]
LGTM

File: gcc/rust/backend/rust-mangle.cc [hunk 1]
```diff
@@ -79,6 +79,92 @@ legacy_mangle_self (const TyTy::BaseType *self)
...
+    return "u";
+
+  // FIXME: ARTHUR: Add rest of algorithm
+  return "";
+}
+
+static std::string
+v0_numeric_prefix (const TyTy::BaseType *ty)
+{
+  static const std::map<std::string, std::string> num_prefixes = {
```
Comment (philberty) [inline gcc/rust/backend/rust-mangle.cc:95]: This is neat :)

File: gcc/rust/backend/rust-mangle.cc [hunk 2]
```diff
@@ -79,6 +79,92 @@ legacy_mangle_self (const TyTy::BaseType *self)
...
+{
+  if (ty->is_unit ())
+    return "u";
+
+  // FIXME: ARTHUR: Add rest of algorithm
+  return "";
+}
+
+static std::string
+v0_numeric_prefix (const TyTy::BaseType *ty)
```
Comment (philberty) [inline gcc/rust/backend/rust-mangle.cc:93]: I think in later PR's you might want to make this more error safe by doing something like:

```
static bool v0_numeric_prefix(const TyTy::BaseType* ty, std::string& prefix)
```

So in future changes you can error by returning false, so it becomes:

```
std::string prefix;
bool ok = v0_numeric_prefix(ty, prefix);
rust_asser(ok);
```

File: gcc/rust/backend/rust-mangle.cc [hunk 3]
```diff
@@ -79,6 +79,92 @@ legacy_mangle_self (const TyTy::BaseType *self)
...
+      return "e";
+    case TyTy::TypeKind::NEVER:
+      return "z";
+
+      // Placeholder types
+    case TyTy::TypeKind::ERROR:	      // Fallthrough
+    case TyTy::TypeKind::INFER:	      // Fallthrough
+    case TyTy::TypeKind::PLACEHOLDER: // Fallthrough
+    case TyTy::TypeKind::PARAM:
+      // FIXME: TyTy::TypeKind::BOUND is also a valid variant in rustc
```
Comment (philberty) [inline gcc/rust/backend/rust-mangle.cc:137]: Good to know thanks for the heads up

**Pull request merged**